src/Grpc.Net.Client.Web/Internal/Base64ResponseStream.cs

```diff
@@ -0,0 +1,276 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.Net.Client.Web.Internal
+{
+    internal class Base64ResponseStream : Stream
+    {
+        private readonly Stream _inner;
+
+        private byte[]? _minimumBuffer;
+        private int _remainder;
+        private byte _remainderByte0;
+        private byte _remainderByte1;
+
+        public Base64ResponseStream(Stream inner)
+        {
+            _inner = inner;
+        }
+
+        public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationToken cancellationToken = default)
+        {
+            // There is enough remaining data to fill passed in data
+            if (data.Length <= _remainder)
+            {
+                return CopyRemainderToData(data);
+            }
+
+            var underlyingReadData = data;
+            var copyFromMinimumBuffer = false;
+            if (data.Length < 6)
+            {
+                // If the requested data is very small, increase it to 6.
+                // 4 bytes for base64, and 2 bytes for potential remaining content.
+                if (_minimumBuffer == null)
+                {
+                    _minimumBuffer = new byte[4 + 2];
+                }
+
+                underlyingReadData = _minimumBuffer;
+                copyFromMinimumBuffer = true;
+            }
+            underlyingReadData = SetRemainder(underlyingReadData);
+
+            // We want to read base64 data in multiples of 4
+            underlyingReadData = underlyingReadData.Slice(0, (underlyingReadData.Length / 4) * 4);
+
+            var availableReadData = underlyingReadData;
+
+            var totalRead = 0;
+            // Minimum valid base64 length is 4. Read until we have at least that much content
+            do
+            {
+                var read = await _inner.ReadAsync(availableReadData, cancellationToken);
```

Comment:
[In Kestrels HttpRequestStream](https://github.com/aspnet/AspNetCore/blob/478353a40feb5f37ef9f1fcea96ba1cf34e1b760/src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestStream.cs#L122) and the [runtime's PipeReaderStream](https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs#L82-L85) (which is used by PipeReader.AsStream()) we throw OperationCanceledExceptions if the underlying pipe was canceled. We should probably do the same here.

Reply:
This code is calling `int ReadAsync(Memory<byte>)` on Stream, not PipeReader. There is no cancel read result